File: Projects/Offensive/Red/scenario_generator.c
/**
 * Red/Blue Exercise Scenarios Generator - C Implementation
 * Compile: gcc -o scenario_generator scenario_generator.c -lm
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <math.h>

#define MAX_OBJECTIVES 10
#define MAX_TECHNIQUES 20
#define MAX_ACTIONS 15
#define MAX_ASSETS 20
#define MAX_IOCS 10
#define MAX_COUNTERMEASURES 8

typedef struct {
    char id[20];
    char name[100];
    char description[500];
    char difficulty[20];
    char mitre_attack_id[20];
    char detection_difficulty[20];
    char iocs[MAX_IOCS][100];
    int ioc_count;
    char countermeasures[MAX_COUNTERMEASURES][100];
    int countermeasure_count;
} AttackTechnique;

typedef struct {
    char name[100];
    char description[500];
    int points;
    char category[50];
} ScenarioObjective;

typedef struct {
    char name[100];
    char description[500];
    char effectiveness[20];
    char tools[200];
    char time_required[50];
} BlueTeamAction;

typedef struct {
    char type[50];
    int count;
    char os[50];
} InfrastructureAsset;

typedef struct {
    char name[200];
    char description[500];
    char difficulty[20];
    char duration[50];
    ScenarioObjective red_team_objectives[MAX_OBJECTIVES];
    int red_obj_count;
    ScenarioObjective blue_team_objectives[MAX_OBJECTIVES];
    int blue_obj_count;
    AttackTechnique attack_techniques[MAX_TECHNIQUES];
    int tech_count;
    BlueTeamAction blue_team_actions[MAX_ACTIONS];
    int action_count;
    InfrastructureAsset assets[MAX_ASSETS];
    int asset_count;
    int user_count;
    char departments[5][50];
    int dept_count;
} ExerciseScenario;

// Predefined attack techniques
AttackTechnique techniques[] = {
    {
        "T1566.001",
        "Phishing - Spearphishing Attachment",
        "Send targeted emails with malicious attachments",
        "Medium",
        "T1566.001",
        "Medium",
        {
            "Unusual email attachments",
            "Suspicious sender domains",
            "Macro-enabled documents",
            "Unusual process spawning"
        },
        4,
        {
            "Email filtering for executables",
            "User awareness training",
            "Endpoint detection for macros",
            "Network monitoring for C2"
        },
        4
    },
    {
        "T1059.003", 
        "Command and Scripting Interpreter - Windows Command Shell",
        "Use cmd.exe for command execution",
        "Low",
        "T1059.003",
        "Easy",
        {
            "Suspicious command-line arguments",
            "Unusual parent-child processes",
            "Commands from unusual locations"
        },
        3,
        {
            "Process monitoring",
            "Application whitelisting",
            "Command-line auditing"
        },
        3
    },
    {
        "T1134",
        "Access Token Manipulation", 
        "Manipulate access tokens to escalate privileges",
        "High",
        "T1134",
        "Hard",
        {
            "Token manipulation API calls",
            "Unusual process integrity levels",
            "Privilege escalation attempts"
        },
        3,
        {
            "Privileged account management",
            "Token filtering",
            "Process integrity monitoring"
        },
        3
    }
};

int technique_count = sizeof(techniques) / sizeof(techniques[0]);

// Scenario templates
typedef struct {
    char name[100];
    char description[500];
    char difficulty[20];
    char duration[50];
    char red_focus[3][50];
    char blue_focus[3][50];
} ScenarioTemplate;

ScenarioTemplate templates[] = {
    {
        "Advanced Persistent Threat Simulation",
        "Simulate sophisticated nation-state actor",
        "High",
        "2 weeks",
        {"Persistence", "Lateral Movement", "Data Exfiltration"},
        {"Threat Hunting", "Incident Response", "Forensics"}
    },
    {
        "Ransomware Attack Scenario",
        "Simulate ransomware deployment and impact", 
        "Medium",
        "3 days",
        {"Initial Access", "Execution", "Impact"},
        {"Containment", "Recovery", "Business Continuity"}
    }
};

int template_count = sizeof(templates) / sizeof(templates[0]);

// Infrastructure templates
typedef struct {
    char type[50];
    char description[200];
    InfrastructureAsset assets[MAX_ASSETS];
    int asset_count;
    int user_count;
    char departments[5][50];
    int dept_count;
} InfrastructureTemplate;

InfrastructureTemplate infrastructures[] = {
    {
        "corporate_network",
        "Standard corporate network environment",
        {
            {"Domain Controller", 2, "Windows Server 2019"},
            {"File Server", 1, "Windows Server 2019"},
            {"Web Server", 2, "Linux Ubuntu"},
            {"Workstation", 20, "Windows 10"},
            {"Network Firewall", 1, ""},
            {"IDS/IPS", 1, ""},
            {"SIEM", 1, ""}
        },
        7,
        50,
        {"HR", "Finance", "IT", "Marketing", "Operations"},
        5
    },
    {
        "cloud_environment",
        "Cloud-native environment",
        {
            {"EC2 Instances", 10, "Linux/Windows"},
            {"S3 Buckets", 5, ""},
            {"RDS Databases", 2, ""},
            {"Load Balancer", 1, ""},
            {"CloudTrail", 1, ""}
        },
        5,
        100,
        {"Development", "Operations", "Security"},
        3
    }
};

int infra_count = sizeof(infrastructures) / sizeof(infrastructures[0]);

void generate_red_objectives(const char* focus_areas[], int focus_count, ScenarioObjective objectives[], int* obj_count) {
    *obj_count = 0;
    
    for(int i = 0; i < focus_count && *obj_count < MAX_OBJECTIVES; i++) {
        if(strcmp(focus_areas[i], "Persistence") == 0) {
            strcpy(objectives[*obj_count].name, "Establish Persistence");
            strcpy(objectives[*obj_count].description, "Create multiple persistence mechanisms");
            objectives[*obj_count].points = 35;
            strcpy(objectives[*obj_count].category, "Persistence");
            (*obj_count)++;
        } else if(strcmp(focus_areas[i], "Lateral Movement") == 0) {
            strcpy(objectives[*obj_count].name, "Domain Compromise");
            strcpy(objectives[*obj_count].description, "Compromise domain administrator account");
            objectives[*obj_count].points = 50;
            strcpy(objectives[*obj_count].category, "Lateral Movement");
            (*obj_count)++;
        } else if(strcmp(focus_areas[i], "Initial Access") == 0) {
            strcpy(objectives[*obj_count].name, "Gain Initial Foothold");
            strcpy(objectives[*obj_count].description, "Establish initial access to target environment");
            objectives[*obj_count].points = 25;
            strcpy(objectives[*obj_count].category, "Initial Access");
            (*obj_count)++;
        }
    }
}

void generate_blue_objectives(const char* focus_areas[], int focus_count, ScenarioObjective objectives[], int* obj_count) {
    *obj_count = 0;
    
    for(int i = 0; i < focus_count && *obj_count < MAX_OBJECTIVES; i++) {
        if(strcmp(focus_areas[i], "Threat Hunting") == 0) {
            strcpy(objectives[*obj_count].name, "IOC Identification");
            strcpy(objectives[*obj_count].description, "Identify multiple IOCs from attack");
            objectives[*obj_count].points = 30;
            strcpy(objectives[*obj_count].category, "Threat Hunting");
            (*obj_count)++;
        } else if(strcmp(focus_areas[i], "Incident Response") == 0) {
            strcpy(objectives[*obj_count].name, "Containment");
            strcpy(objectives[*obj_count].description, "Contain the incident within 2 hours");
            objectives[*obj_count].points = 35;
            strcpy(objectives[*obj_count].category, "Incident Response");
            (*obj_count)++;
        }
    }
}

void select_techniques(const char* difficulty, AttackTechnique selected_techs[], int* tech_count) {
    *tech_count = 0;
    
    for(int i = 0; i < technique_count && *tech_count < MAX_TECHNIQUES; i++) {
        if(strcmp(difficulty, "Low") == 0 && strcmp(techniques[i].difficulty, "Low") == 0) {
            selected_techs[*tech_count] = techniques[i];
            (*tech_count)++;
        } else if(strcmp(difficulty, "Medium") == 0 && 
                 (strcmp(techniques[i].difficulty, "Low") == 0 || 
                  strcmp(techniques[i].difficulty, "Medium") == 0)) {
            selected_techs[*tech_count] = techniques[i];
            (*tech_count)++;
        } else if(strcmp(difficulty, "High") == 0) {
            selected_techs[*tech_count] = techniques[i];
            (*tech_count)++;
        }
    }
}

void generate_blue_actions(AttackTechnique techs[], int tech_count, BlueTeamAction actions[], int* action_count) {
    *action_count = 0;
    
    for(int i = 0; i < tech_count && *action_count < MAX_ACTIONS; i++) {
        for(int j = 0; j < techs[i].countermeasure_count && *action_count < MAX_ACTIONS; j++) {
            snprintf(actions[*action_count].name, sizeof(actions[*action_count].name), 
                    "Defend against %s", techs[i].name);
            snprintf(actions[*action_count].description, sizeof(actions[*action_count].description),
                    "Implement %s", techs[i].countermeasures[j]);
            strcpy(actions[*action_count].effectiveness, "High");
            strcpy(actions[*action_count].tools, "SIEM, EDR, Network Monitoring");
            strcpy(actions[*action_count].time_required, "1-2 hours");
            (*action_count)++;
        }
    }
}

void generate_scenario(const char* scenario_type, const char* infrastructure_type, 
                      const char* difficulty, ExerciseScenario* scenario) {
    // Find template
    ScenarioTemplate* template = NULL;
    for(int i = 0; i < template_count; i++) {
        if(strstr(templates[i].name, scenario_type) != NULL) {
            template = &templates[i];
            break;
        }
    }
    if(!template) template = &templates[0];
    
    // Find infrastructure
    InfrastructureTemplate* infra = NULL;
    for(int i = 0; i < infra_count; i++) {
        if(strcmp(infrastructures[i].type, infrastructure_type) == 0) {
            infra = &infrastructures[i];
            break;
        }
    }
    if(!infra) infra = &infrastructures[0];
    
    // Set basic scenario info
    snprintf(scenario->name, sizeof(scenario->name), "%s - %s", template->name, infra->description);
    strcpy(scenario->description, template->description);
    strcpy(scenario->difficulty, difficulty);
    strcpy(scenario->duration, template->duration);
    
    // Generate objectives
    generate_red_objectives(template->red_focus, 3, scenario->red_team_objectives, &scenario->red_obj_count);
    generate_blue_objectives(template->blue_focus, 3, scenario->blue_team_objectives, &scenario->blue_obj_count);
    
    // Select techniques
    select_techniques(difficulty, scenario->attack_techniques, &scenario->tech_count);
    
    // Generate blue team actions
    generate_blue_actions(scenario->attack_techniques, scenario->tech_count, 
                         scenario->blue_team_actions, &scenario->action_count);
    
    // Copy infrastructure
    scenario->asset_count = infra->asset_count;
    for(int i = 0; i < infra->asset_count; i++) {
        scenario->assets[i] = infra->assets[i];
    }
    scenario->user_count = infra->user_count;
    scenario->dept_count = infra->dept_count;
    for(int i = 0; i < infra->dept_count; i++) {
        strcpy(scenario->departments[i], infra->departments[i]);
    }
}

void print_scenario_json(const ExerciseScenario* scenario) {
    printf("{\n");
    printf("  \"name\": \"%s\",\n", scenario->name);
    printf("  \"description\": \"%s\",\n", scenario->description);
    printf("  \"difficulty\": \"%s\",\n", scenario->difficulty);
    printf("  \"duration\": \"%s\",\n", scenario->duration);
    
    printf("  \"red_team_objectives\": [\n");
    for(int i = 0; i < scenario->red_obj_count; i++) {
        printf("    {\n");
        printf("      \"name\": \"%s\",\n", scenario->red_team_objectives[i].name);
        printf("      \"description\": \"%s\",\n", scenario->red_team_objectives[i].description);
        printf("      \"points\": %d,\n", scenario->red_team_objectives[i].points);
        printf("      \"category\": \"%s\"\n", scenario->red_team_objectives[i].category);
        printf("    }%s\n", i < scenario->red_obj_count - 1 ? "," : "");
    }
    printf("  ],\n");
    
    printf("  \"infrastructure\": {\n");
    printf("    \"assets\": [\n");
    for(int i = 0; i < scenario->asset_count; i++) {
        printf("      {\n");
        printf("        \"type\": \"%s\",\n", scenario->assets[i].type);
        printf("        \"count\": %d,\n", scenario->assets[i].count);
        printf("        \"os\": \"%s\"\n", scenario->assets[i].os);
        printf("      }%s\n", i < scenario->asset_count - 1 ? "," : "");
    }
    printf("    ],\n");
    printf("    \"users\": %d\n", scenario->user_count);
    printf("  }\n");
    printf("}\n");
}

void print_scenario_text(const ExerciseScenario* scenario) {
    printf("=== RED/BLUE EXERCISE SCENARIO ===\n\n");
    printf("Name: %s\n", scenario->name);
    printf("Description: %s\n", scenario->description);
    printf("Difficulty: %s\n", scenario->difficulty);
    printf("Duration: %s\n\n", scenario->duration);
    
    printf("RED TEAM OBJECTIVES:\n");
    for(int i = 0; i < scenario->red_obj_count; i++) {
        printf("  %d. %s (%d points)\n", i+1, scenario->red_team_objectives[i].name, 
               scenario->red_team_objectives[i].points);
        printf("     %s\n", scenario->red_team_objectives[i].description);
    }
    
    printf("\nBLUE TEAM OBJECTIVES:\n");
    for(int i = 0; i < scenario->blue_obj_count; i++) {
        printf("  %d. %s (%d points)\n", i+1, scenario->blue_team_objectives[i].name,
               scenario->blue_team_objectives[i].points);
        printf("     %s\n", scenario->blue_team_objectives[i].description);
    }
    
    printf("\nINFRASTRUCTURE:\n");
    for(int i = 0; i < scenario->asset_count; i++) {
        printf("  - %s: %d", scenario->assets[i].type, scenario->assets[i].count);
        if(strlen(scenario->assets[i].os) > 0) {
            printf(" (%s)", scenario->assets[i].os);
        }
        printf("\n");
    }
    printf("  Total Users: %d\n", scenario->user_count);
}

int main(int argc, char* argv[]) {
    char* scenario_type = "apt_simulation";
    char* infrastructure = "corporate_network";
    char* difficulty = "Medium";
    char* format = "text";
    
    // Simple argument parsing
    for(int i = 1; i < argc; i++) {
        if(strcmp(argv[i], "--scenario-type") == 0 && i+1 < argc) {
            scenario_type = argv[++i];
        } else if(strcmp(argv[i], "--infrastructure") == 0 && i+1 < argc) {
            infrastructure = argv[++i];
        } else if(strcmp(argv[i], "--difficulty") == 0 && i+1 < argc) {
            difficulty = argv[++i];
        } else if(strcmp(argv[i], "--format") == 0 && i+1 < argc) {
            format = argv[++i];
        }
    }
    
    srand(time(NULL));
    
    ExerciseScenario scenario;
    generate_scenario(scenario_type, infrastructure, difficulty, &scenario);
    
    if(strcmp(format, "json") == 0) {
        print_scenario_json(&scenario);
    } else {
        print_scenario_text(&scenario);
    }
    
    return 0;
}

File: Projects/Offensive/Red/scenario_generator.py
#!/usr/bin/env python3
"""
Red/Blue Exercise Scenarios Generator - Python Implementation
Creates realistic cyber exercise scenarios for training
"""

import random
import json
import yaml
import argparse
import datetime
from dataclasses import dataclass, asdict
from typing import List, Dict, Any
import hashlib

@dataclass
class AttackTechnique:
    id: str
    name: str
    description: str
    difficulty: str  # Low, Medium, High
    mitre_attack_id: str
    detection_difficulty: str  # Easy, Medium, Hard
    ioc_indicators: List[str]
    countermeasures: List[str]

@dataclass
class ScenarioObjective:
    name: str
    description: str
    points: int
    category: str  # InitialAccess, Execution, Persistence, etc.

@dataclass
class BlueTeamAction:
    name: str
    description: str
    effectiveness: str  # High, Medium, Low
    tools: List[str]
    time_required: str  # Minutes, Hours, Days

@dataclass
class ExerciseScenario:
    name: str
    description: str
    difficulty: str
    duration: str
    red_team_objectives: List[ScenarioObjective]
    blue_team_objectives: List[ScenarioObjective]
    attack_techniques: List[AttackTechnique]
    blue_team_actions: List[BlueTeamAction]
    infrastructure: Dict[str, Any]
    scoring_rubric: Dict[str, Any]

class ScenarioGenerator:
    def __init__(self):
        self.attack_techniques = self._load_attack_techniques()
        self.infrastructure_templates = self._load_infrastructure_templates()
        self.scenario_templates = self._load_scenario_templates()
        
    def _load_attack_techniques(self) -> List[AttackTechnique]:
        """Load predefined attack techniques"""
        return [
            AttackTechnique(
                id="T1566.001",
                name="Phishing - Spearphishing Attachment",
                description="Send targeted emails with malicious attachments",
                difficulty="Medium",
                mitre_attack_id="T1566.001",
                detection_difficulty="Medium",
                ioc_indicators=[
                    "Unusual email attachments",
                    "Suspicious sender domains", 
                    "Macro-enabled documents",
                    "Unusual process spawning from email clients"
                ],
                countermeasures=[
                    "Email filtering for executable attachments",
                    "User awareness training",
                    "Endpoint detection for macro execution",
                    "Network monitoring for C2 communications"
                ]
            ),
            AttackTechnique(
                id="T1059.003",
                name="Command and Scripting Interpreter - Windows Command Shell",
                description="Use cmd.exe for command execution",
                difficulty="Low",
                mitre_attack_id="T1059.003",
                detection_difficulty="Easy",
                ioc_indicators=[
                    "Suspicious command-line arguments",
                    "Unusual parent-child process relationships",
                    "Commands from unusual locations",
                    "Base64 encoded commands"
                ],
                countermeasures=[
                    "Process monitoring and logging",
                    "Application whitelisting",
                    "Command-line argument auditing",
                    "EDR solutions"
                ]
            ),
            AttackTechnique(
                id="T1134",
                name="Access Token Manipulation",
                description="Manipulate access tokens to escalate privileges",
                difficulty="High",
                mitre_attack_id="T1134",
                detection_difficulty="Hard",
                ioc_indicators=[
                    "Token manipulation API calls",
                    "Unusual process integrity levels",
                    "Privilege escalation attempts",
                    "Abnormal token assignments"
                ],
                countermeasures=[
                    "Privileged account management",
                    "Token filtering",
                    "Process integrity monitoring",
                    "Least privilege principle"
                ]
            ),
            AttackTechnique(
                id="T1027",
                name="Obfuscated Files or Information",
                description="Use obfuscation to hide malicious code",
                difficulty="Medium",
                mitre_attack_id="T1027",
                detection_difficulty="Hard",
                ioc_indicators=[
                    "Encrypted or encoded network traffic",
                    "Suspicious file entropy",
                    "Obfuscated PowerShell scripts",
                    "Packed executables"
                ],
                countermeasures=[
                    "Network traffic analysis",
                    "File entropy analysis",
                    "Behavioral detection",
                    "Memory analysis"
                ]
            )
        ]
    
    def _load_infrastructure_templates(self) -> Dict[str, Any]:
        """Load infrastructure templates for different scenarios"""
        return {
            "corporate_network": {
                "description": "Standard corporate network environment",
                "assets": [
                    {"type": "Domain Controller", "count": 2, "os": "Windows Server 2019"},
                    {"type": "File Server", "count": 1, "os": "Windows Server 2019"},
                    {"type": "Web Server", "count": 2, "os": "Linux Ubuntu"},
                    {"type": "Workstation", "count": 20, "os": "Windows 10"},
                    {"type": "Network Firewall", "count": 1},
                    {"type": "IDS/IPS", "count": 1},
                    {"type": "SIEM", "count": 1}
                ],
                "users": 50,
                "departments": ["HR", "Finance", "IT", "Marketing", "Operations"]
            },
            "industrial_control": {
                "description": "Industrial Control System (ICS) environment",
                "assets": [
                    {"type": "HMI", "count": 3, "os": "Windows 10"},
                    {"type": "PLC", "count": 10},
                    {"type": "Engineering Workstation", "count": 5, "os": "Windows 10"},
                    {"type": "Historian", "count": 1, "os": "Windows Server 2019"},
                    {"type": "ICS Firewall", "count": 2},
                    {"type": "SCADA Server", "count": 2, "os": "Windows Server 2019"}
                ],
                "users": 25,
                "departments": ["Operations", "Engineering", "Maintenance"]
            },
            "cloud_environment": {
                "description": "Cloud-native environment",
                "assets": [
                    {"type": "EC2 Instances", "count": 10, "os": "Linux/Windows"},
                    {"type": "S3 Buckets", "count": 5},
                    {"type": "RDS Databases", "count": 2},
                    {"type": "Load Balancer", "count": 1},
                    {"type": "CloudTrail", "count": 1},
                    {"type": "CloudWatch", "count": 1},
                    {"type": "WAF", "count": 1}
                ],
                "users": 100,
                "departments": ["Development", "Operations", "Security"]
            }
        }
    
    def _load_scenario_templates(self) -> Dict[str, Any]:
        """Load scenario templates"""
        return {
            "apt_simulation": {
                "name": "Advanced Persistent Threat Simulation",
                "description": "Simulate sophisticated nation-state actor",
                "difficulty": "High",
                "duration": "2 weeks",
                "red_focus": ["Persistence", "Lateral Movement", "Data Exfiltration"],
                "blue_focus": ["Threat Hunting", "Incident Response", "Forensics"]
            },
            "ransomware_attack": {
                "name": "Ransomware Attack Scenario",
                "description": "Simulate ransomware deployment and impact",
                "difficulty": "Medium", 
                "duration": "3 days",
                "red_focus": ["Initial Access", "Execution", "Impact"],
                "blue_focus": ["Containment", "Recovery", "Business Continuity"]
            },
            "insider_threat": {
                "name": "Insider Threat Scenario", 
                "description": "Simulate malicious insider activities",
                "difficulty": "Medium",
                "duration": "1 week",
                "red_focus": ["Privilege Abuse", "Data Theft", "Sabotage"],
                "blue_focus": ["User Monitoring", "DLP", "Behavioral Analysis"]
            },
            "supply_chain": {
                "name": "Supply Chain Compromise",
                "description": "Simulate attack through third-party vendor",
                "difficulty": "High",
                "duration": "10 days", 
                "red_focus": ["Initial Access", "Trust Abuse", "Lateral Movement"],
                "blue_focus": ["Vendor Risk Management", "Network Segmentation", "Monitoring"]
            }
        }
    
    def generate_scenario(self, scenario_type: str, infrastructure: str, difficulty: str = "Medium") -> ExerciseScenario:
        """Generate a complete exercise scenario"""
        template = self.scenario_templates.get(scenario_type, self.scenario_templates["apt_simulation"])
        infra = self.infrastructure_templates.get(infrastructure, self.infrastructure_templates["corporate_network"])
        
        # Select appropriate attack techniques based on difficulty
        techniques = self._select_techniques(difficulty)
        
        # Generate objectives
        red_objectives = self._generate_red_team_objectives(template["red_focus"])
        blue_objectives = self._generate_blue_team_objectives(template["blue_focus"])
        
        # Generate blue team actions
        blue_actions = self._generate_blue_team_actions(techniques)
        
        # Generate scoring rubric
        scoring = self._generate_scoring_rubric(red_objectives, blue_objectives)
        
        return ExerciseScenario(
            name=f"{template['name']} - {infra['description']}",
            description=template["description"],
            difficulty=difficulty,
            duration=template["duration"],
            red_team_objectives=red_objectives,
            blue_team_objectives=blue_objectives,
            attack_techniques=techniques,
            blue_team_actions=blue_actions,
            infrastructure=infra,
            scoring_rubric=scoring
        )
    
    def _select_techniques(self, difficulty: str) -> List[AttackTechnique]:
        """Select attack techniques based on difficulty"""
        difficulty_map = {
            "Low": ["Low"],
            "Medium": ["Low", "Medium"],
            "High": ["Low", "Medium", "High"]
        }
        
        allowed_difficulties = difficulty_map.get(difficulty, ["Medium"])
        return [tech for tech in self.attack_techniques if tech.difficulty in allowed_difficulties]
    
    def _generate_red_team_objectives(self, focus_areas: List[str]) -> List[ScenarioObjective]:
        """Generate red team objectives"""
        objectives = []
        
        objective_templates = {
            "InitialAccess": [
                ("Gain Initial Foothold", "Establish initial access to the target environment", 25),
                ("Phishing Campaign", "Successfully deliver phishing payload to multiple users", 30)
            ],
            "Persistence": [
                ("Establish Persistence", "Create at least 2 persistence mechanisms", 35),
                ("Backdoor Installation", "Install backdoor on critical system", 40)
            ],
            "LateralMovement": [
                ("Domain Compromise", "Compromise domain administrator account", 50),
                ("Network Enumeration", "Map the entire network infrastructure", 25)
            ],
            "DataExfiltration": [
                ("Data Collection", "Identify and collect sensitive data", 30),
                ("Exfiltration", "Successfully exfiltrate data without detection", 45)
            ],
            "Execution": [
                ("Code Execution", "Execute arbitrary code on multiple systems", 20),
                ("Privilege Escalation", "Escalate privileges on critical system", 35)
            ]
        }
        
        for area in focus_areas:
            if area in objective_templates:
                for name, desc, points in objective_templates[area]:
                    objectives.append(ScenarioObjective(
                        name=name,
                        description=desc,
                        points=points,
                        category=area
                    ))
        
        return objectives
    
    def _generate_blue_team_objectives(self, focus_areas: List[str]) -> List[ScenarioObjective]:
        """Generate blue team objectives"""
        objectives = []
        
        objective_templates = {
            "Threat Hunting": [
                ("IOC Identification", "Identify at least 5 IOCs from attack", 30),
                ("Attack Timeline", "Reconstruct complete attack timeline", 40)
            ],
            "Incident Response": [
                ("Containment", "Contain the incident within 2 hours", 35),
                ("Eradication", "Remove all attacker persistence mechanisms", 45)
            ],
            "Forensics": [
                ("Evidence Collection", "Collect forensic evidence from 3 systems", 25),
                ("Root Cause Analysis", "Determine root cause of compromise", 30)
            ],
            "Monitoring": [
                ("Alert Tuning", "Create new detection rules based on attack", 20),
                ("False Positive Reduction", "Reduce false positives by 50%", 25)
            ]
        }
        
        for area in focus_areas:
            if area in objective_templates:
                for name, desc, points in objective_templates[area]:
                    objectives.append(ScenarioObjective(
                        name=name,
                        description=desc,
                        points=points,
                        category=area
                    ))
        
        return objectives
    
    def _generate_blue_team_actions(self, techniques: List[AttackTechnique]) -> List[BlueTeamAction]:
        """Generate recommended blue team actions"""
        actions = []
        
        for tech in techniques:
            for countermeasure in tech.countermeasures:
                actions.append(BlueTeamAction(
                    name=f"Defend against {tech.name}",
                    description=f"Implement {countermeasure}",
                    effectiveness="High",
                    tools=["SIEM", "EDR", "Network Monitoring"],
                    time_required="1-2 hours"
                ))
        
        # Add general blue team actions
        general_actions = [
            BlueTeamAction(
                name="Network Traffic Analysis",
                description="Analyze network flows for suspicious patterns",
                effectiveness="Medium",
                tools=["Wireshark", "Zeek", "Suricata"],
                time_required="2-4 hours"
            ),
            BlueTeamAction(
                name="Endpoint Investigation",
                description="Conduct deep dive on compromised endpoints",
                effectiveness="High", 
                tools=["EDR", "Volatility", "Audit Logs"],
                time_required="3-5 hours"
            ),
            BlueTeamAction(
                name="User Awareness Assessment",
                description="Evaluate user susceptibility to social engineering",
                effectiveness="Medium",
                tools=["Phishing Simulation", "Training Platforms"],
                time_required="1 day"
            )
        ]
        
        actions.extend(general_actions)
        return actions
    
    def _generate_scoring_rubric(self, red_objs: List[ScenarioObjective], blue_objs: List[ScenarioObjective]) -> Dict[str, Any]:
        """Generate scoring rubric for the exercise"""
        return {
            "red_team_scoring": {
                "objective_completion": {
                    "description": "Points for completed objectives",
                    "max_points": sum(obj.points for obj in red_objs)
                },
                "stealth": {
                    "description": "Points for avoiding detection",
                    "max_points": 50
                },
                "time_efficiency": {
                    "description": "Points for completing objectives quickly",
                    "max_points": 30
                }
            },
            "blue_team_scoring": {
                "objective_completion": {
                    "description": "Points for completed objectives", 
                    "max_points": sum(obj.points for obj in blue_objs)
                },
                "detection_time": {
                    "description": "Points for quick detection",
                    "max_points": 40
                },
                "containment_effectiveness": {
                    "description": "Points for effective containment",
                    "max_points": 35
                },
                "documentation": {
                    "description": "Points for thorough documentation",
                    "max_points": 25
                }
            },
            "win_conditions": {
                "red_team_win": "Score 70% of total possible points",
                "blue_team_win": "Score 70% of total possible points and contain attack within 4 hours",
                "draw": "Both teams score between 50-70% of possible points"
            }
        }
    
    def generate_inject(self, scenario: ExerciseScenario, inject_type: str) -> Dict[str, Any]:
        """Generate exercise injects (events during the exercise)"""
        inject_templates = {
            "phishing_email": {
                "type": "Phishing Email",
                "description": "Simulated phishing email sent to employees",
                "difficulty": "Medium",
                "iocs": ["Suspicious sender", "Malicious attachment", "Urgent language"],
                "response_actions": ["Analyze email headers", "Check attachment hashes", "Block sender"]
            },
            "malware_detection": {
                "type": "Malware Alert",
                "description": "Endpoint detection alert for suspicious process",
                "difficulty": "High", 
                "iocs": ["Unknown process", "Suspicious network connections", "File modifications"],
                "response_actions": ["Isolate endpoint", "Collect memory dump", "Analyze process tree"]
            },
            "privilege_escalation": {
                "type": "Privilege Escalation Attempt",
                "description": "Attempt to gain higher privileges on system",
                "difficulty": "Hard",
                "iocs": ["UAC bypass attempts", "Token manipulation", "Service installation"],
                "response_actions": ["Review privilege logs", "Check service configurations", "Analyze authentication events"]
            }
        }
        
        template = inject_templates.get(inject_type, inject_templates["phishing_email"])
        return {
            "id": hashlib.md5(f"{scenario.name}{datetime.datetime.now()}".encode()).hexdigest()[:8],
            "scenario": scenario.name,
            "timestamp": datetime.datetime.now().isoformat(),
            **template
        }
    
    def export_scenario(self, scenario: ExerciseScenario, format: str = "json") -> str:
        """Export scenario in specified format"""
        data = asdict(scenario)
        
        if format.lower() == "json":
            return json.dumps(data, indent=2)
        elif format.lower() == "yaml":
            return yaml.dump(data, default_flow_style=False)
        else:
            return str(data)

def main():
    parser = argparse.ArgumentParser(description="Red/Blue Exercise Scenarios Generator")
    parser.add_argument("--scenario-type", choices=["apt_simulation", "ransomware_attack", "insider_threat", "supply_chain"],
                       default="apt_simulation", help="Type of scenario to generate")
    parser.add_argument("--infrastructure", choices=["corporate_network", "industrial_control", "cloud_environment"],
                       default="corporate_network", help="Infrastructure type")
    parser.add_argument("--difficulty", choices=["Low", "Medium", "High"], default="Medium",
                       help="Exercise difficulty level")
    parser.add_argument("--format", choices=["json", "yaml", "text"], default="json",
                       help="Output format")
    parser.add_argument("--inject", action="store_true", help="Generate sample inject")
    
    args = parser.parse_args()
    
    generator = ScenarioGenerator()
    scenario = generator.generate_scenario(args.scenario_type, args.infrastructure, args.difficulty)
    
    if args.inject:
        inject = generator.generate_inject(scenario, "phishing_email")
        print(f"Generated Inject:\n{json.dumps(inject, indent=2)}")
    else:
        output = generator.export_scenario(scenario, args.format)
        print(output)

if __name__ == "__main__":
    main()
